dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/program/notification/ProgramNotificationServiceIntegrationTest.java

```diff
@@ -0,0 +1,159 @@
+import org.hisp.dhis.user.User;
+import org.hisp.dhis.user.UserGroup;
+import org.hisp.dhis.user.UserService;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.Timeout;
+import org.springframework.beans.factory.annotation.Autowired;
+
+/**
+ * @author Zubair Asghar
+ */
+class ProgramNotificationServiceIntegrationTest extends SingleSetupIntegrationTestBase {
+  private static final String PI_UID = "PI_UID";
+  private static final int TEST_USER_COUNT = 30;
+
+  @Autowired private IdentifiableObjectManager manager;
+  @Autowired private ProgramNotificationService programNotificationService;
+  @Autowired private ProgramInstanceService programInstanceService;
+  @Autowired private ProgramRuleEngineService programRuleEngineService;
+  @Autowired private ProgramRuleService programRuleService;
+  @Autowired private ProgramRuleActionService programRuleActionService;
+  @Autowired protected UserService _userService;
+
+  private Program program;
+  private OrganisationUnit organisationUnit;
+  private ProgramRule programRule;
+  private ProgramInstance programInstance;
+  private TrackedEntityInstance trackedEntityInstance;
+  private UserGroup userGroup;
+
+  @Override
+  public void setUpTest() throws ParseException {
+    userService = _userService;
+    organisationUnit = createOrganisationUnit('O');
+    manager.save(organisationUnit);
+    program = createProgram('P', Set.of(), organisationUnit);
+    manager.save(program);
+    programRule = createProgramRule('R', program);
+    programRule.setCondition("true");
+    manager.save(programRule);
+
+    userGroup = createUserGroup('U', new HashSet<>());
+    manager.save(userGroup);
+
+    trackedEntityInstance = createTrackedEntityInstance('T', organisationUnit);
+    manager.save(trackedEntityInstance);
+    programInstance = createProgramInstance(program, trackedEntityInstance, organisationUnit);
+    programInstance.setUid(PI_UID);
+    manager.save(programInstance);
+
+    createAndAddUsersToUserGroupAndOrgUnit();
+
+    setUpScheduleMessage("V{current_date}", 'A');
+    setUpScheduleMessage("V{current_date}", 'B');
+    setUpScheduleMessage("V{current_date}", 'C');
+    setUpScheduleMessage("'2025-01-01'", 'E');
+
+    ProgramInstance pi = programInstanceService.getProgramInstance(PI_UID);
+    List<RuleEffect> ruleEffects =
+        programRuleEngineService.evaluateEnrollmentAndRunEffects(pi.getId());
+    assertEquals(
+        4, ruleEffects.size(), "Expected 3 rule effects for scheduled messages " + ruleEffects);
+  }
+
+  @Test
+  @Timeout(value = 30, unit = TimeUnit.SECONDS)
+  void testScheduledNotificationShouldBeTriggeredBeforeTimeout() {
+    /* No assertions are made here because this test is intended to verify that
+    scheduled notifications are fetched and processed within the specified time limit
+    (performance check only) */
+    programNotificationService.sendScheduledNotifications(NoopJobProgress.INSTANCE);
```

Comment:
This test is not really needed is it? The `notificationInstanceService.getAll()` was the issue right? So if you add a test to the `ProgramNotificationInstanceServiceTest` for `getAll` and the method you are using now showing they do not timeout that would be enough. This would also allow you to assert on their results.

This test here could be useful but IMHO you would need to assert that the notifications you expected to be sent were actually sent.